src/cbc_sdk/platform/grants.py

```diff
@@ -485,6 +485,30 @@ def _delete_object(self):
         ret = self._cb.delete_object(url)
         self._refresh_if_needed(ret)
 
+    @classmethod
+    def get_permitted_roles(cls, cb):
+        """
+        Returns a list of all permitted roles that we can assign to a user.
+
+        Args:
+            cb (CBCloudAPI): A reference to the CBCloudAPI object.
+
+        Returns:
+            set: A set of string role URNs that we are permitted to manage (assign to users).
+        """
+        def _extractor(resp_data):
+            """Pulls out all the lists of URN descriptors from the result."""
+            for d in resp_data.values():
+                for ll in d.values():
+                    yield ll
+
+        token_split = cb.credentials.token.split('/')
+        url = "/access/v3/orgs/{0}/principals/{1}/roles/permitted?type=USER".format(cb.credentials.org_key,
```

Comment:
If we are sending the API_ID should `type=API_KEY`? Is it possible that I could get the permitted roles for a user by it's user_id while authenticating with an API_KEY?

Reply:
Or does the type filter the roles by Access Level ones vs User roles?

Reply:
If we can get the permitted roles for a User using an API_Key then we should consider adding this as a function to the User object as well

Reply:
We can't get permitted roles for an arbitrary user. What we _can_ get are the set of roles that _you,_ the person making the request, are permitted to add to users.  Using USER as the type parameter returns this.  Using API_KEY returns the set of roles that you are permitted to add to API keys. This is not useful to us at this point.

Reply:
To clarify, the `principal_id` in the API path has to match to the id of the principal that is authenticating otherwise we get a 403 Forbidden?

Reply:
Yes. To be more precise, the principal_id in the path _has_ to be the "token" portion of the value you've supplied as the X-Auth-Token. So everything after the / character.  Otherwise, it's 403 Forbidden.